src/rpp/rpp/operators/tap.hpp

```diff
@@ -0,0 +1,158 @@
+//                  ReactivePlusPlus library
+//
+//          Copyright Aleksey Loginov 2023 - present.
+// Distributed under the Boost Software License, Version 1.0.
+//    (See accompanying file LICENSE_1_0.txt or copy at
+//          https://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/victimsnino/ReactivePlusPlus
+//
+
+#pragma once
+
+#include <rpp/operators/fwd.hpp>
+
+#include <rpp/defs.hpp>
+#include <rpp/operators/details/strategy.hpp>
+
+namespace rpp::operators::details
+{
+template<
+    rpp::constraint::observer     TObserver,
+    rpp::constraint::decayed_type OnNext,
+    rpp::constraint::decayed_type OnError,
+    rpp::constraint::decayed_type OnCompleted>
+struct tap_observer_strategy
+{
+    using preferred_disposable_strategy = rpp::details::observers::none_disposable_strategy;
+
+    RPP_NO_UNIQUE_ADDRESS TObserver   observer;
+    RPP_NO_UNIQUE_ADDRESS OnNext      onNext;
+    RPP_NO_UNIQUE_ADDRESS OnError     onError;
+    RPP_NO_UNIQUE_ADDRESS OnCompleted onCompleted;
+
+    template<typename T>
+    void on_next(T&& v) const
+    {
+        onNext(utils::as_const(v));
+        observer.on_next(std::forward<T>(v));
+    }
+
+    void on_error(const std::exception_ptr& err) const
+    {
+        onError(err);
+        observer.on_error(err);
+    }
+
+    void on_completed() const
+    {
+        onCompleted();
+        observer.on_completed();
+    }
+
+    void set_upstream(const disposable_wrapper& d) { observer.set_upstream(d); }
+
+    bool is_disposed() const { return observer.is_disposed(); }
+};
+
+template<
+    rpp::constraint::decayed_type OnNext,
+    rpp::constraint::decayed_type OnError,
+    rpp::constraint::decayed_type OnCompleted>
+struct tap_t : public operators::details::operator_observable_strategy<tap_observer_strategy, OnNext, OnError, OnCompleted>
+{
+    using operators::details::operator_observable_strategy<tap_observer_strategy, OnNext, OnError, OnCompleted>::operator_observable_strategy;
+
+    template<rpp::constraint::decayed_type T>
+        requires rpp::constraint::invocable_r_v<void, OnNext, T>
+    using result_value = T;
+
+    template<rpp::details::observables::constraint::disposable_strategy Prev>
+    using updated_disposable_strategy = Prev;
+};
+}
+
+namespace rpp::operators
+{
+/**
+ * @brief Register callbacks to inspect observable emissions and perform side-effects
+ *
+ * @param on_error error handler
+ *
+ * @ingroup utility_operators
+ * @see https://reactivex.io/documentation/operators/do.html
+ */
+template<typename OnNext                                       = rpp::utils::empty_function_any_t,
+         std::invocable<const std::exception_ptr&> OnError     = rpp::utils::empty_function_t<std::exception_ptr>,
+         std::invocable<>                          OnCompleted = rpp::utils::empty_function_t<>>
```

Comment:
OnNext/OnCompleted should be omitted in this case and passed directly to details_tap_t

Reply:
done